File: scripts/normalize_pdf.py
import sys
import os
from pypdf import PdfReader, PdfWriter

def normalize(input_path, output_path):
    try:
        reader = PdfReader(input_path)
        writer = PdfWriter()

        for page in reader.pages:
            writer.add_page(page)

        # Write the normalized PDF
        with open(output_path, "wb") as f:
            writer.write(f)

        # Verify file creation
        if os.path.exists(output_path) and os.path.getsize(output_path) > 0:
            print("Success")
            sys.exit(0)
        else:
            print("Error: Output file not created or empty")
            sys.exit(1)

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python normalize_pdf.py <input> <output>")
        sys.exit(1)

    normalize(sys.argv[1], sys.argv[2])

File: scripts/normalize_pdf.js
const { PDFDocument } = require('pdf-lib');
const fs = require('fs').promises;

async function normalizePdf() {
    const args = process.argv.slice(2);
    if (args.length < 2) {
        console.error('Usage: node normalize_pdf.js <input_path> <output_path>');
        process.exit(1);
    }

    const inputPath = args[0];
    const outputPath = args[1];

    try {
        const pdfBytes = await fs.readFile(inputPath);

        // Load the PDF document
        const pdfDoc = await PDFDocument.load(pdfBytes);

        // Save the PDF with useObjectStreams: false to maximize compatibility (v1.4-like structure)
        const pdfBytesNormalized = await pdfDoc.save({ useObjectStreams: false });

        await fs.writeFile(outputPath, pdfBytesNormalized);

        console.log('PDF normalized successfully');
        process.exit(0);
    } catch (error) {
        console.error('Error normalizing PDF:', error);
        process.exit(1);
    }
}

normalizePdf();
